File: trunk/java/picocontainer/src/test/org/picocontainer/defaults/ImplementationHidingComponentAdapterFactoryTestCase.java
/*****************************************************************************
 * Copyright (ComponentC) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/

package org.picocontainer.defaults;

import org.picocontainer.MutablePicoContainer;
import org.picocontainer.tck.AbstractComponentAdapterFactoryTestCase;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class ImplementationHidingComponentAdapterFactoryTestCase extends AbstractComponentAdapterFactoryTestCase {
    private ImplementationHidingComponentAdapterFactory implementationHiddingComponentAdapterFactory = new ImplementationHidingComponentAdapterFactory(new DefaultComponentAdapterFactory());
    private CachingComponentAdapterFactory cachingComponentAdapterFactory = new CachingComponentAdapterFactory(implementationHiddingComponentAdapterFactory);

    public static interface Man {
        Woman getWoman();

        void kiss();

        boolean wasKissed();
    }

    public static interface Woman {
        Man getMan();
    }

    public static interface SuperWoman extends Woman {
    }

    public static class Husband implements Man {
        public final Woman partner;
        private boolean wasKissed;

        public Husband(Woman partner) {
            this.partner = partner;
        }

        public Woman getWoman() {
            return partner;
        }

        public void kiss() {
            wasKissed = true;
        }

        public boolean wasKissed() {
            return wasKissed;
        }
    }

    public static class Wife implements SuperWoman {
        public final Man partner;

        public Wife(Man partner) {
            this.partner = partner;
        }

        public Man getMan() {
            return partner;
        }
    }

    public void testLowLevelCheating() {
        ComponentAdapterFactory caf = createComponentAdapterFactory();
        DefaultPicoContainer pico = new DefaultPicoContainer(caf);

        CachingComponentAdapter wifeAdapter = (CachingComponentAdapter) caf.createComponentAdapter("wife", Wife.class, null);
        CachingComponentAdapter husbandAdapter = (CachingComponentAdapter) caf.createComponentAdapter("husband", Husband.class, null);

        pico.registerComponent(wifeAdapter);
        pico.registerComponent(husbandAdapter);

        Woman wife = (Woman) wifeAdapter.getComponentInstance();
        wife.getMan().kiss();
        Man man = (Man) husbandAdapter.getComponentInstance();
        assertTrue(man.wasKissed());

        assertSame(man, wife.getMan());
        assertSame(wife, man.getWoman());

        // Let the wife use another (single) man
        Man newMan = new Husband(null);
        Man oldMan = (Man) ((Swappable) man).__hotSwap(newMan);

        wife.getMan().kiss();
        assertTrue(newMan.wasKissed());
        assertNotSame(man, oldMan);
        assertNotSame(oldMan, newMan);
        assertNotSame(newMan, man);

        assertFalse(man.hashCode() == oldMan.hashCode());
        assertFalse(oldMan.hashCode() == newMan.hashCode());
        assertFalse(newMan.hashCode() == man.hashCode());
    }

    public void testHighLevelCheating() {
        MutablePicoContainer pico = new DefaultPicoContainer(createComponentAdapterFactory());

        // Register two classes with mutual dependencies in the constructor (!!!)
        pico.registerComponentImplementation(Wife.class);
        pico.registerComponentImplementation(Husband.class);

        Woman wife = (Woman) pico.getComponentInstance(Wife.class);
        Man man = (Man) pico.getComponentInstance(Husband.class);

        assertSame(man, wife.getMan());
        assertSame(wife, man.getWoman());

        // Let the wife use another (single) man
        Man newMan = new Husband(null);
        Man oldMan = (Man) ((Swappable) man).__hotSwap(newMan);

        wife.getMan().kiss();
        assertFalse(oldMan.wasKissed());
        assertTrue(newMan.wasKissed());

    }

    public void testBigamy() {
        DefaultPicoContainer pico = new DefaultPicoContainer(new ImplementationHidingComponentAdapterFactory(
                new ConstructorComponentAdapterFactory()));
        pico.registerComponentImplementation(Woman.class, Wife.class);
        Woman firstWife = (Woman) pico.getComponentInstance(Woman.class);
        Woman secondWife = (Woman) pico.getComponentInstance(Woman.class);
        assertNotSame(firstWife, secondWife);

    }

    //TODO we need to be able to control which interfaces are subject to dynamic proxy
    // we used to have this feature with registerComponentImplementationByType or somesuch.

    public void do_nottestProxiedWifeIsNotSuperWoman() {
        DefaultPicoContainer pico = new DefaultPicoContainer(new ImplementationHidingComponentAdapterFactory(
                new ConstructorComponentAdapterFactory()));
        pico.registerComponentImplementation(Woman.class, Wife.class);
        Woman wife = (Woman) pico.getComponentInstance(Woman.class);
        assertFalse("Wife should not be castable to SuperWoman, as she should be proxied to Woman only", wife instanceof SuperWoman);
    }


    public static class Bad implements Serializable {
        public Bad() {
            throw new IllegalStateException("HAHA");
        }
    }

    public void testIHCAFwithCTORandNoCaching() {
        // http://lists.codehaus.org/pipermail/picocontainer-dev/2004-January/001985.html
        MutablePicoContainer pico = new DefaultPicoContainer();
        pico.registerComponent(new ImplementationHidingComponentAdapter(new ConstructorComponentAdapter("l", ArrayList.class)));

        List list1 = (List) pico.getComponentInstance("l");
        List list2 = (List) pico.getComponentInstance("l");

        assertNotSame(list1, list2);

        list1.add("Hello");
        assertTrue(list1.contains("Hello"));
        assertFalse(list2.contains("Hello"));
    }

    public void testSwappingViaSwappableInterface() {
        MutablePicoContainer pico = new DefaultPicoContainer();
        pico.registerComponent(new ImplementationHidingComponentAdapter(new ConstructorComponentAdapter("l", ArrayList.class)));
        List l = (List) pico.getComponentInstance("l");
        l.add("Hello");
        final ArrayList newList = new ArrayList();
        ArrayList oldSubject = (ArrayList) ((Swappable)l).__hotSwap(newList);
        assertEquals("Hello", oldSubject.get(0));
        assertTrue(l.isEmpty());
        l.add("World");
        assertEquals("World", l.get(0));
    }

    public interface OtherSwappable {
        Object __hotSwap(Object newSubject);
    }

    public static class OtherSwappableImpl implements OtherSwappable {
        public Object __hotSwap(Object newSubject) {
            return "TADA";
        }
    }

    public void testInterferingSwapMethodsInComponentMasksHotSwappingFunctionality() {
        MutablePicoContainer pico = new DefaultPicoContainer();
        pico.registerComponent(new ImplementationHidingComponentAdapter(new ConstructorComponentAdapter("os", OtherSwappableImpl.class)));
        OtherSwappable os = (OtherSwappable) pico.getComponentInstance("os");
        OtherSwappable os2 = new OtherSwappableImpl();

        assertEquals("TADA", os.__hotSwap(os2));
        Swappable os_ = (Swappable) os;
        assertEquals("TADA", os_.__hotSwap(os2));
    }

    protected ComponentAdapterFactory createComponentAdapterFactory() {
        return cachingComponentAdapterFactory;
    }
}

File: trunk/dotnet/picocontainer/src/Test/Defaults/ImplementationHidingComponentAdapterFactoryTestCase.cs
using System;
using System.Collections;
using PicoContainer;
using PicoContainer.Defaults;
using PicoContainer.Tests.Tck;
using PicoContainer.Tests.TestModel;

using NUnit.Framework;

namespace Test.Defaults {
  /// <summary>
  /// Summary description for ImplementationHidingComponentAdapterFactoryTestCase.
  /// </summary>
  [TestFixture]
  public class ImplementationHidingComponentAdapterFactoryTestCase {// : AbstractComponentAdapterFactoryTestCase {
    private static ImplementationHidingComponentAdapterFactory implementationHiddingComponentAdapterFactory = new ImplementationHidingComponentAdapterFactory(new DefaultComponentAdapterFactory());
    private CachingComponentAdapterFactory cachingComponentAdapterFactory = new CachingComponentAdapterFactory(implementationHiddingComponentAdapterFactory);

    public  interface Man {
      Woman getWoman();

      void kiss();

      bool wasKissed();
    }

    public interface Woman {
      Man getMan();
    }

    public interface SuperWoman : Woman {
    }

    public class Husband : Man {
      public readonly Woman partner;
      private bool _wasKissed;

      public Husband(Woman partner) {
        this.partner = partner;
      }

      public Woman getWoman() {
        return partner;
      }

      public void kiss() {
        _wasKissed = true;
      }

      public bool wasKissed() {
        return _wasKissed;
      }
    }

    public class Wife : SuperWoman {
      public readonly Man partner;

      public Wife(Man partner) {
        this.partner = partner;
      }

      public Man getMan() {
        return partner;
      }
    }

    [Test]
    public void testLowLevelCheating() {
      IComponentAdapterFactory caf = CreateComponentAdapterFactory();
      DefaultPicoContainer pico = new DefaultPicoContainer(caf);

      CachingComponentAdapter wifeAdapter = (CachingComponentAdapter) caf.CreateComponentAdapter("wife", typeof(Wife), null);
      CachingComponentAdapter husbandAdapter = (CachingComponentAdapter) caf.CreateComponentAdapter("husband", typeof(Husband), null);

      pico.RegisterComponent(wifeAdapter);
      pico.RegisterComponent(husbandAdapter);

      Woman wife = (Woman) wifeAdapter.ComponentInstance;
      wife.getMan().kiss();
      Man man = (Man) husbandAdapter.ComponentInstance;
      Assert.IsTrue(man.wasKissed());

      Assert.AreSame(man, wife.getMan());
      Assert.AreSame(wife, man.getWoman());

      // Let the wife use another (single) man
      Man newMan = new Husband(null);
      Man oldMan = (Man) ((ISwappable) man).HotSwap(newMan);

      Assert.IsFalse(newMan.wasKissed());
      wife.getMan().kiss();
      Assert.IsTrue(newMan.wasKissed());
      Assert.IsFalse(man.Equals(oldMan));
      Assert.IsFalse(oldMan.Equals(newMan));
      Assert.IsFalse(newMan.Equals(man));

      Assert.IsFalse(man.GetHashCode() == oldMan.GetHashCode());
      Assert.IsFalse(oldMan.GetHashCode() == newMan.GetHashCode());
      // C# implementation fails, there is for as far as I know no System.identityHashCode available for c#
      //      Assert.IsFalse(newMan.GetHashCode() == man.GetHashCode());
    }

    
    [Test]
    public void testHighLevelCheating() {
      IMutablePicoContainer pico = new DefaultPicoContainer(CreateComponentAdapterFactory());

      // Register two classes with mutual dependencies in the constructor (!!!)
      pico.RegisterComponentImplementation(typeof(Wife));
      pico.RegisterComponentImplementation(typeof(Husband));

      Woman wife = (Woman) pico.GetComponentInstance(typeof(Wife));
      Man man = (Man) pico.GetComponentInstance(typeof(Husband));

      Assert.AreSame(man, wife.getMan());
      Assert.AreSame(wife, man.getWoman());

      // Let the wife use another (single) man
      Man newMan = new Husband(null);
      Man oldMan = (Man) ((ISwappable) man).HotSwap(newMan);

      wife.getMan().kiss();
      Assert.IsFalse(oldMan.wasKissed());
      Assert.IsTrue(newMan.wasKissed());

    }


    [Test]
    public void testBigamy() {
      DefaultPicoContainer pico = new DefaultPicoContainer(new ImplementationHidingComponentAdapterFactory(new ConstructorInjectionComponentAdapterFactory()));
      pico.RegisterComponentImplementation(typeof(Woman), typeof(Wife));
      Woman firstWife = (Woman) pico.GetComponentInstance(typeof(Woman));
      Woman secondWife = (Woman) pico.GetComponentInstance(typeof(Woman));
      Assert.IsFalse(firstWife == secondWife);

    }


    [Test]
    public void testComponentsRegisteredWithClassKeyOnlyImplementThatInterface() {
      DefaultPicoContainer pico = new DefaultPicoContainer(new ImplementationHidingComponentAdapterFactory(new ConstructorInjectionComponentAdapterFactory()));
      pico.RegisterComponentImplementation(typeof(Woman), typeof(Wife));
      Woman wife = (Woman) pico.GetComponentInstance(typeof(Woman));
      Assert.IsFalse(wife is SuperWoman);
    }


    public void testIHCAFwithCTORandNoCaching() {
      // http://lists.codehaus.org/pipermail/picocontainer-dev/2004-January/001985.html
      IMutablePicoContainer pico = new DefaultPicoContainer();
      pico.RegisterComponent(new ImplementationHidingComponentAdapter(new ConstructorInjectionComponentAdapter("l", typeof(ArrayList))));

      IList list1 = (IList) pico.GetComponentInstance("l");
      IList list2 = (IList) pico.GetComponentInstance("l");

      Assert.IsFalse(list1 == list2);
      Assert.IsFalse(list1 is ArrayList);

      list1.Add("Hello");
      Assert.IsTrue(list1.Contains("Hello"));
      Assert.IsFalse(list2.Contains("Hello"));
    }

    public void testSwappingViaSwappableInterface() {
      IMutablePicoContainer pico = new DefaultPicoContainer();
      pico.RegisterComponent(new ImplementationHidingComponentAdapter(new ConstructorInjectionComponentAdapter("l", typeof(ArrayList))));
      IList l = (IList) pico.GetComponentInstance("l");
      l.Add("Hello");
      ArrayList newList = new ArrayList();
      ArrayList oldSubject = (ArrayList) ((ISwappable) l).HotSwap(newList);
      Assert.AreEqual("Hello", oldSubject[0]);
      Assert.IsTrue(l.Count == 0);
      l.Add("World");
      Assert.AreEqual("World", l[0]);
    }


    public interface OtherSwappable {
      object HotSwap(object newSubject);
    }

    public class OtherSwappableImpl : OtherSwappable {
      public virtual object HotSwap(object newSubject) {
        return "TADA";
      }
    }


    public void testInterferingSwapMethodsInComponentMasksHotSwappingFunctionality() {
      IMutablePicoContainer pico = new DefaultPicoContainer();
      pico.RegisterComponent(new ImplementationHidingComponentAdapter(new ConstructorInjectionComponentAdapter("os", typeof(OtherSwappableImpl))));
      OtherSwappable os = (OtherSwappable) pico.GetComponentInstance("os");
      OtherSwappable os2 = new OtherSwappableImpl();

      Assert.AreEqual("TADA", os.HotSwap(os2));
      ISwappable os_ = (ISwappable) os;
      Assert.AreEqual("TADA", os_.HotSwap(os2));
    }



    protected IComponentAdapterFactory CreateComponentAdapterFactory() {
      return cachingComponentAdapterFactory;
    }


  }
}
